File: examples/plugins/c/noise/drifting_noise.c
// Example propaq native noise plugin, implemented in C.
//
// A demonstration of *stateful ctx*: damping grows with the number of calls
// already made through this ctx, rather than staying fixed like
// UniformNoiseModel.
//
//   damping_factor = exp(-damping * weight * (1 + drift_rate * call_index))
//
// NOTE: This model is not layer-dependent depolarizing noise. 
//

#include <math.h>
#include <stdatomic.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#define PROPAQ_NOISE_ABI_VERSION 1u

typedef struct {
    double damping;
    double drift_rate;
    _Atomic uint64_t call_index;
} Ctx;

uint32_t propaq_noise_abi_version(void) {
    return PROPAQ_NOISE_ABI_VERSION;
}

static double parse_double(const char* config_json, const char* key, double fallback) {
    if (config_json == NULL) return fallback;
    const char* found = strstr(config_json, key);
    if (found == NULL) return fallback;
    const char* colon = strchr(found, ':');
    if (colon == NULL) return fallback;
    return atof(colon + 1);
}

void* propaq_noise_create(const char* config_json) {
    Ctx* ctx = (Ctx*)malloc(sizeof(Ctx));
    ctx->damping = parse_double(config_json, "\"damping\"", 0.001);
    ctx->drift_rate = parse_double(config_json, "\"drift_rate\"", 0.0);
    atomic_init(&ctx->call_index, 0);
    return ctx;
}

void propaq_noise_destroy(void* ctx) {
    free(ctx);
}

static double damping_factor(Ctx* ctx, uint32_t term_weight, uint64_t call_index) {
    double factor = 1.0 + ctx->drift_rate * (double)call_index;
    return exp(-ctx->damping * (double)term_weight * factor);
}

double propaq_noise_factor(void* ctx, uint32_t basis_kind, const uint64_t* words, size_t n_words,
                           uint32_t n_units, uint32_t weight, uint32_t layer_index,
                           uint32_t n_layers) {
    (void)basis_kind; (void)words; (void)n_words; (void)n_units;
    (void)layer_index; (void)n_layers;
    Ctx* c = (Ctx*)ctx;
    uint64_t call_index = atomic_fetch_add_explicit(&c->call_index, 1, memory_order_relaxed);
    return damping_factor(c, weight, call_index);
}

int32_t propaq_noise_factor_batch(void* ctx, uint32_t basis_kind, const uint64_t* words,
                                  size_t n_words_per_term, uint32_t n_units,
                                  const uint32_t* weights, uint32_t layer_index, uint32_t n_layers,
                                  double* out, size_t n_terms) {
    (void)basis_kind; (void)words; (void)n_words_per_term; (void)n_units;
    (void)layer_index; (void)n_layers;
    Ctx* c = (Ctx*)ctx;
    uint64_t base = atomic_fetch_add_explicit(&c->call_index, (uint64_t)n_terms, memory_order_relaxed);
    for (size_t i = 0; i < n_terms; i++) {
        out[i] = damping_factor(c, weights[i], base + (uint64_t)i);
    }
    return 0;
}

File: examples/plugins/julia/noise/drifting_noise.jl
# Example propaq noise plugin, implemented in Julia (AOT-compiled via
# PackageCompiler.jl into a C-ABI shared library; see the plugins README).
#
# A demonstration of *stateful ctx*: damping grows with the number of calls
# already made through this ctx, rather than staying fixed like
# UniformNoiseModel.
#
#   damping_factor = exp(-damping * weight * (1 + drift_rate * call_index))
#
# This is not depth-dependent noise.
# 

const PROPAQ_NOISE_ABI_VERSION = UInt32(1)

mutable struct Ctx
    damping::Float64
    drift_rate::Float64
    call_index::Threads.Atomic{UInt64}
end

const CTX_STORE = Dict{Ptr{Cvoid},Ctx}()
const CTX_LOCK = ReentrantLock()

function parse_field(json::AbstractString, key::AbstractString, fallback::Float64)
    m = match(Regex("\"" * key * "\"\\s*:\\s*(-?[0-9.eE+-]+)"), json)
    m === nothing ? fallback : parse(Float64, m.captures[1])
end

function damping_factor(damping::Float64, drift_rate::Float64, weight::UInt32, call_index::UInt64)
    factor = 1.0 + drift_rate * Float64(call_index)
    return exp(-damping * Float64(weight) * factor)
end

Base.@ccallable function propaq_noise_abi_version()::UInt32
    PROPAQ_NOISE_ABI_VERSION
end

Base.@ccallable function propaq_noise_create(config_json::Ptr{Cchar})::Ptr{Cvoid}
    json = config_json == C_NULL ? "" : unsafe_string(config_json)
    ctx = Ctx(parse_field(json, "damping", 0.001), parse_field(json, "drift_rate", 1e-5), Threads.Atomic{UInt64}(0))
    key = pointer_from_objref(ctx)
    lock(CTX_LOCK) do
        CTX_STORE[key] = ctx
    end
    return key
end

Base.@ccallable function propaq_noise_destroy(ctx::Ptr{Cvoid})::Cvoid
    lock(CTX_LOCK) do
        delete!(CTX_STORE, ctx)
    end
    return nothing
end

Base.@ccallable function propaq_noise_factor(ctx::Ptr{Cvoid}, basis_kind::UInt32,
                                             words::Ptr{UInt64}, n_words::Csize_t,
                                             n_units::UInt32, weight::UInt32,
                                             layer_index::UInt32, n_layers::UInt32)::Cdouble
    c = unsafe_pointer_to_objref(ctx)::Ctx
    call_index = Threads.atomic_add!(c.call_index, UInt64(1))
    damping_factor(c.damping, c.drift_rate, weight, call_index)
end

Base.@ccallable function propaq_noise_factor_batch(ctx::Ptr{Cvoid}, basis_kind::UInt32,
                                                   words::Ptr{UInt64}, n_words_per_term::Csize_t,
                                                   n_units::UInt32, weights::Ptr{UInt32},
                                                   layer_index::UInt32, n_layers::UInt32,
                                                   out::Ptr{Cdouble}, n_terms::Csize_t)::Int32
    c = unsafe_pointer_to_objref(ctx)::Ctx
    # Reserve the whole chunk's worth of call indices with one atomic op
    # instead of one fetch-add per term.
    base = Threads.atomic_add!(c.call_index, UInt64(n_terms))
    w = unsafe_wrap(Array, weights, n_terms)
    result = unsafe_wrap(Array, out, n_terms)
    @inbounds for i in 1:n_terms
        result[i] = damping_factor(c.damping, c.drift_rate, w[i], base + UInt64(i - 1))
    end
    return Int32(0)
end
